### ISSUE TITLE
Nested messages do not encode from JSON to GRPC properly

### ISSUE THREAD

**Issue by kgoguevgoget** [ysugimoto/lua-resty-grpc-gateway#9]
Hi, I think there is an issue with messages containing other types of messages when you send the request in to be transformed from JSON to GRPC. 

If we modify the example helloworld.proto and attempt to call the server with the following curl request

```
curl -vv -H "Content-Type: application/json" -d '{"name":"grpc-rest", "ex":[{"displayname":"test","grades":[1,2,3]}], "jobs":["A","B"]}' "http://localhost:9000/rest"
```

```
syntax = "proto3";
package helloworld;

import "google/protobuf/timestamp.proto";

service Greeter {
  rpc SayHello (HelloRequest) returns (HelloReply) {}
}

message HelloRequest {
  string name = 1;
  repeated ComplexMsg ex =2; /** Example of nested message type**/
  repeated string jobs =3;
}

message ComplexMsg {
  string displayname = 1;
  repeated int32 grades = 2;
}

message HelloReply {
  string message = 1;
  google.protobuf.Timestamp reply_at = 2;
}
```
The request does get transformed but the grpc request object in Go does not get filled correctly.
Mainly the array objects are empty.

```go
(*helloworld.HelloRequest)(0xc00005c780)(name:"grpc-rest" jobs:"A" jobs:"B" )
```
Looking though the code, I think the issue is in the `_M.map_message = function(field, default_values)` in the `util.lua`.  But I do not have enough lua knowledge or knowledge of how the requests are transformed to submit a patch. 

This is a small dump of the request that is being built by the map_message function. I understand that it is a recursive function when it encounters field_types starting with `.` but the request that comes back from that is `{}`

```
 [lua] util.lua:73: map_message(): name: name field_type: string, 
 [lua] util.lua:73: map_message(): name: ex field_type: .helloworld.ComplexMsg, 
 [lua] util.lua:73: map_message(): name: displayname field_type: string, 
 [lua] util.lua:73: map_message(): name: grades field_type: int32, 
 [lua] util.lua:84: map_message(): Request: {}, 
 [lua] util.lua:73: map_message(): name: jobs field_type: string, 
 [lua] util.lua:84: map_message(): Request: {
   ex = {},
   jobs = { "A", "B" },
   name = "grpc-rest"
 }, 

```

**Comment by kgoguevgoget** [comment]
Further investigation, it seems like the code recursively searches the `pb.fields` it finds the inner most field and value but because the message contains a nested message ( for example
"ex":[{"displayname":"test","grades":[1,2,3]}), it fails to set the `default_values` or the `sub` variable. 

```
sub, err = _M.map_message(field_type, default_values)
      if err then
        return nil, err
      end
      request[name] = sub
```
```
request[name] = get_from_request(name, field_type) or default_values[name] or nil
-- request[displayname] = nil or nil or nil  
--- request[jobs] = nil or { "A", "B" } or nil
```

**Comment by kgoguevgoget** [comment]
This is a possible fix, it seems like the problem pertains to the lua request table

given a table of the following 
```
{ name = "grpc-rest",
  ex = { displayname = "test"},
  jobs = { "A", "B"}
}
```
the `request[name]` will only work on flat structures, when it comes to the nested object, trying to use
`request[name]` like so `request["displayname"]` will return nil you need to call `request["ex"]` first so the table that is returned is the inner object and then the `request["displayname"]` will work.

I have attempted to make the recursive function use the inner objects

```
_M.map_message = function(field, default_values)
  if not pb.type(field) then
    return nil, ("Field %s is not defined"):format(field)
  end
  local request = {}

  for name, _, field_type in pb.fields(field) do
    if field_type:sub(1, 1) == "." then
      default_values = json.decode(ngx.req.get_body_data())  -- Extract the values from the data and then set the table with the content of the key 
      sub, err = _M.map_message(field_type, default_values[name])  -- the request[name] expects a flat key object structure but this is the not the case when we have nested objects..
      if err then
        return nil, err
      end
      request[name] = sub
    else
      request[name] = get_from_request(name, field_type) or default_values[name] or nil
    end
  end
  return request, nil
end
```

**Comment by ysugimoto** [comment]
@kgoguevgoget Sorry for the late reply and thank you for your investigation. 
As you write, I was lack of consideration that nested message.. so I'll look into it and try to fix it.

**Comment by kgoguevgoget** [comment]
Hi no worries about the timing, I will actually submit a pull request  to you in an hour or two Swedish time. I was interested in trying to solve it. I think I got it working actually... though feel free to totally destroy my patch as I don't understand Lua very well (Sorry I am a Go developer)

**Comment by ysugimoto** [comment]
@kgoguevgoget That's great! Welcome your PR!
And I would say no worries because I'm a Go developer, too 😅

**Comment by kgoguevgoget** [comment]
Hi again the following PR #10 should work.. have a look, sorry about the delay it was more than an hour  or two .

**Comment by ysugimoto** [comment]
@kgoguevgoget No problem, I'll look at it on tomorrow, thanks!

**Comment by ysugimoto** [comment]
@kgoguevgoget Hi, I looked you PR and approved with a tiny suggestion -- almost looks fine!

Well, I'll merge it after you confirmed my suggestion and please let me add you as a contributor if you don't mind.

**Comment by kgoguevgoget** [comment]
@ysugimoto Excellent, I've made the suggested change, please double check that you can still run it and get the expected results .. I am not sure my dev environment is setup properly for lua and openresty. 

Of course I would be happy if you added me as a contributor.

**Comment by kgoguevgoget** [comment]
@ysugimoto Thanks for merging this, however it looks like the circleci now requires luajit as part of the container that is building and running the tests.. this is due to the use of `table.isempty()`

**Comment by ysugimoto** [comment]
@kgoguevgoget Yes, finally I've fixed that https://github.com/ysugimoto/lua-resty-grpc-gateway/pull/11

in local, we use `penlight` as you wrote first 👍 because our CI is running with lua, not luajit.

And, I've just sent an invitation as a contributor to this project. please check it when you have time, thanks again!

**Comment by kgoguevgoget** [comment]
@ysugimoto Ah ok that makes sense... Wouldn't it be easier to run with LuaJIT anyway? anyway the publish step of the circleci seems to lack `penlight`

**Comment by ysugimoto** [comment]
@kgoguevgoget I think that using https://github.com/GUI/lua-docker/tree/master/luajit-2.1-beta on CI base image would be better, so I'll try it later.

**Comment by kgoguevgoget** [comment]
@ysugimoto Looks like the latest version (v1.2.x) was built now but didn't get uploaded to luarocks

**Comment by ysugimoto** [comment]
@kgoguevgoget Sorry, I've just published to luarocks as v1.2.2

**Comment by kgoguevgoget** [comment]
@ysugimoto Great this works now.
